core/trie/proof.go

```diff
@@ -54,623 +55,586 @@ func (e *Edge) Len() uint8 {
-// ordering of the merged path is not important
-// since SplitProofPath can discover the left and right paths using the merged path and the rootHash
-func MergeProofPaths(leftPath, rightPath []ProofNode, hash hashFunc) ([]ProofNode, *felt.Felt, error) {
-	merged := []ProofNode{}
-	minLen := min(len(leftPath), len(rightPath))
-
-	if len(leftPath) == 0 || len(rightPath) == 0 {
-		return merged, nil, errors.New("empty proof paths")
-	}
-
-	if !leftPath[0].Hash(hash).Equal(rightPath[0].Hash(hash)) {
-		return merged, nil, errors.New("roots of the proof paths are different")
+// VerifyRangeProof checks whether the given key-value pairs and range proof are valid for the given root hash.
+// The range of key-value pairs should be consecutive (no gap inside) and monotonic increasing.
+// The given range proof contains 2 edge proofs. One is for the first key and the other is for the last key.
+// Both of them can be existent or non-existent.
+// This function also handles the following special cases:
+//
+//   - All elements proof: In this case the proof can be nil, but the range should be all the leaves in the trie.
+//   - One element proof: In this case, both left and right edge proofs are the same and the range should contain only one element.
+//   - Zero element proof: In this case a single edge proof is enough to verify. If there are still more elements, the proof is invalid.
+//
+// If the range proof is invalid, this function returns false and an error.
+// Otherwise, it will return a flag to indicate if there are more elements.
+//
+// TODO(weiihann): Given a binary leaf and a left-sibling first key, if the right sibling is removed, the proof would still be valid.
+// Conversely, given a binary leaf and a right-sibling last key, if the left sibling is removed, the proof would still be valid.
+// Range proof should not be valid for both of these cases, but currently is, which is an attack vector.
+// The problem probably lies in how we do root hash calculation.
+func VerifyRangeProof(root, first *felt.Felt, keys, values []*felt.Felt, proof *ProofNodeSet) (bool, error) { //nolint:funlen,gocyclo
+	// Ensure the number of keys and values are the same
+	if len(keys) != len(values) {
+		return false, fmt.Errorf("inconsistent proof data, keys: %d, values: %d", len(keys), len(values))
 	}
 
-	rootHash := leftPath[0].Hash(hash)
-
-	// Get duplicates and insert by one
-	i := 0
-	for i = 0; i < minLen; i++ {
-		leftNode := leftPath[i]
-		rightNode := rightPath[i]
+	// Ensure all keys are monotonic increasing and values contain no deletions
+	for i := 0; i < len(keys); i++ {
+		if i < len(keys)-1 && keys[i].Cmp(keys[i+1]) > 0 {
+			return false, errors.New("keys are not monotonic increasing")
+		}
 
-		if leftNode.Hash(hash).Equal(rightNode.Hash(hash)) {
-			merged = append(merged, leftNode)
-		} else {
-			break
+		if values[i] == nil || values[i].Equal(&felt.Zero) {
+			return false, errors.New("range contains deletion")
```

Comment:
```suggestion
			return false, errors.New("range contains empty leaf")
```
Yes, might be a result of the deletion, but let's be more straight